src/main/java/ay2122s1_cs2103t_w16_2/btbb/model/predicate/PredicateCollection.java

```diff
@@ -0,0 +1,122 @@
+
+    private boolean hasSamePredicates(List<Predicate<T>> otherPredicates) {
+        return predicates.containsAll(otherPredicates) && otherPredicates.containsAll(predicates);
+    }
+
+    /**
+     * Adds a predicate to the list of predicates to test against.
+     *
+     * @param predicate The Predicate to add to the list.
+     */
+    public void addPredicate(Predicate<T> predicate) {
+        predicates.add(predicate);
+    }
+
+    /**
+     * Adds a {@code ValueEqualsKeywordsPredicate} for quantity values to the list of predicates.
+     *
+     * @param prefix Prefix of keyword.
+     * @param argMultimap ArgumentMultimap to get the value associated with a prefix.
+     * @param getter Function to get the quantity to be tested.
+     * @throws ParseException if the given keywords is invalid.
+     */
+    public void addQuantityEqualsKeywordsPredicate(Prefix prefix, ArgumentMultimap argMultimap,
+            Function<T, Quantity> getter) throws ParseException {
+        if (argMultimap.getValue(prefix).isPresent()) {
+            addPredicate(new ValueEqualsKeywordsPredicate<>(getter,
+                    ParserUtil.parseQuantityKeywords(argMultimap.getValue(prefix).get())));
+        }
+    }
+
+    /**
+     * Adds a {@code ValueWithinRangePredicate} for quantity values to the list of predicates.
+     * It adds the predicate as long as either the lower or upper bound of the range is provided,
+     * and fills the unprovided bound with a default value.
+     *
+     * @param fromPrefix Prefix for the lower bound of the range.
+     * @param toPrefix Prefix for the upper bound of the range.
+     * @param argMultimap ArgumentMultimap to get the value associated with a prefix.
+     * @param getter Function to get the quantity to be tested.
+     * @throws ParseException if the given keywords is invalid.
+     */
+    public void addQuantityWithinRangePredicate(Prefix fromPrefix, Prefix toPrefix,
+            ArgumentMultimap argMultimap, Function<T, Quantity> getter) throws ParseException {
+        Optional<String> optionalMinQuantityKeyword = argMultimap.getValue(fromPrefix);
+        Optional<String> optionalMaxQuantityKeyword = argMultimap.getValue(toPrefix);
+
+        if (optionalMinQuantityKeyword.isEmpty() && optionalMaxQuantityKeyword.isEmpty()) {
+            return;
+        }
+
+        String defaultKeywordMinQuantity = String.valueOf(Integer.MIN_VALUE);
+        String defaultKeywordMaxQuantity = String.valueOf(Integer.MAX_VALUE);
+
+        String minQuantityKeyword = optionalMinQuantityKeyword.orElse(defaultKeywordMinQuantity);
+        String maxQuantityKeyword = optionalMaxQuantityKeyword.orElse(defaultKeywordMaxQuantity);
+
+        addPredicate(new ValueWithinRangePredicate<>(
+                getter,
+                ParserUtil.parseInternalQuantity(minQuantityKeyword),
+                ParserUtil.parseInternalQuantity(maxQuantityKeyword)));
+    }
```

Comment:
I think that this class is being too tightly coupled to quantity. By right it should be able to add a ValueEqualsKeywordsPredicate and ValueWithinRangePredicate for any field type. If you look at the addStringContainsKeywordsPredicate method, it is able to work with all kinds of fields. I'm also not very sure how to do this.

Reply:
Do you mean these methods are too tightly coupled to quantity? I think for each type (quantity, date, price...) we need a separate method because they have slightly different requirements before they can be passed into the general predicate. It's similar to how each type has a different parseX method.